### PR TITLE
GridFieldSetup Automation

### PR DESCRIPTION
**Pull request by johnnywilkes**
## Contributing to Cortex XSOAR Content
Make sure to register your contribution by filling the [contribution registration form](https://forms.gle/XDfxU4E61ZwEESSMA)

**The Pull Request will be reviewed only after the contribution registration form is filled.**

## Status
- [x] In Progress
- [ ] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
fixes: link to the issue

## Description
Grid fields can be super helpful when you want to keep track of entries, but don't know how many there will be.  For instance, maybe you want to keep track of all the system IDs associated with a cloud asset.  Depending on what you keep track of, that would be 1 sys ID or it could be 100.  Therefore, a gridfield can be used in that situation to keep track of all those sys IDs with ease.

One challenge with grid fields, is when using the !setGridField command, the information inputed needs to be in the same context location.  I created !GridFieldSetup to create temp context entry (list of dictionaries) that can be used to pass into !setGridField.  This makes it so you can pull in context from different areas (or static entries) and add to a grid easily.  An example (see test playbook): lets say we have some context populated already
image

We can then use this command to pull them into a single context entry:
`!GridFieldSetup context_path="Temp" keys="IPs,Cloud" val1=${IPs} val2=${Clouds}`
image

This can then be used by a !setGridField command to populate the grid field:
`!SetGridField columns=IP,Source context_path=Temp grid_id=asmprivateip keys=IPs,Cloud`
image
image

 

## Screenshots
Paste here any images that will help the reviewer

## Minimum version of Cortex XSOAR
- [ ] 6.0.0
- [ ] 6.1.0
- [ ] 6.2.0
- [X] 6.5.0

## Does it break backward compatibility?
   - [ ] Yes
       - Further details:
   - [X] No

## Must have
- [X] Tests
- [X] Documentation

### CODE REVIEW

**Comment by johnnywilkes** [comment]
@altmannyarden , this is the automation we talked about to setup context to send to grid field if you are interested

**Comment by johnnywilkes** [comment]
image
I am not sure how these got associated with this PR.  Any ideas how to remove them?

**Comment by johnnywilkes** [comment]
@esharf , thanks for helping me get that sorted.  Looks good to me!

File: Packs/CommonScripts/Scripts/GridFieldSetup/GridFieldSetup.py [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+    if keys:
+        key_list = keys.split(',')
+    num_entries = len(key_list)
+    res_list = []
+    temp = {}
+    count = 0
+    while count < num_entries:
+        tkey = 'val' + str(count + 1)
+        temp[key_list[count]] = vals[tkey]
+        count += 1
```
Comment (esharf) [inline Packs/CommonScripts/Scripts/GridFieldSetup/GridFieldSetup.py:26]:
```suggestion
        count += 1
        temp[key_list[count]] = vals[f'val{count}']
```

File: Packs/CommonScripts/Scripts/GridFieldSetup/GridFieldSetup.py [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+    :type keys: ``str``
+    :type vals: ``dict``
+    :param keys: comma separated values used for keys (columns)
+    :param vals: dictionary of the value assigned to keys
+    :return: list of dictionaries
+    """
+    if keys:
+        key_list = keys.split(',')
+    num_entries = len(key_list)
+    res_list = []
```
Comment (esharf) [inline Packs/CommonScripts/Scripts/GridFieldSetup/GridFieldSetup.py:20]: why do we need this as a list?

**Review by esharf — CHANGES_REQUESTED** [review]
Nice job.
check also this failure [here](https://app.circleci.com/pipelines/github/demisto/content/199446/workflows/1de75d2a-0ea8-4afe-a628-28a92c4e3af5/jobs/584412?invite=true#step-108-94).

**Comment by johnnywilkes** [comment]
> Nice job. check also this failure [here](https://app.circleci.com/pipelines/github/demisto/content/199446/workflows/1de75d2a-0ea8-4afe-a628-28a92c4e3af5/jobs/584412?invite=true#step-108-94).

fixed in [updated docker](https://github.com/demisto/content/pull/21641/commits/7c70abe4e09052539be30c3cc9fe9ed48e11be7b)

File: Packs/CommonScripts/Scripts/GridFieldSetup/GridFieldSetup.py [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+def grid_field_setup(keys: Optional[Any], vals: dict) -> List[Dict[str, Any]]:
+    """Returns a list of dictionaries based on the key/values provided.
+    :type keys: ``str``
+    :type vals: ``dict``
+    :param keys: comma separated values used for keys (columns)
+    :param vals: dictionary of the value assigned to keys
+    :return: list of dictionaries
+    """
+    if keys:
+        key_list = keys.split(',')
```
[the remark this answers is not in the document]
Comment (johnnywilkes) [reply Packs/CommonScripts/Scripts/GridFieldSetup/GridFieldSetup.py]: already fixed in commit
[switch to argToList](https://github.com/demisto/content/pull/21641/commits/3c29bbb2b836a8b3243fbeb9368bedea4d44399b)

File: Packs/CommonScripts/Scripts/GridFieldSetup/GridFieldSetup.py [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+def grid_field_setup_command(args: Dict[str, Any]) -> CommandResults:
+    keys = args.get('keys')
+    context_path = args.get('context_path')
+    # list of all val keys
+    valkeys = sorted([k for k in args.keys() if k.startswith('val')])
+
+    # dictionary or all vals
+    vals = {}
+    for i, val in enumerate(valkeys, start=1):
+        vals[f'val{i}'] = args.get(f'val{i}')
```
[the remark this answers is not in the document]
Comment (johnnywilkes) [reply Packs/CommonScripts/Scripts/GridFieldSetup/GridFieldSetup.py:44]: @esharf , What is the difference between these two?  Spaces vs tabs?

File: Packs/CommonScripts/Scripts/GridFieldSetup/GridFieldSetup.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (johnnywilkes) [reply Packs/CommonScripts/Scripts/GridFieldSetup/GridFieldSetup.py]: This change has already been made

File: Packs/CommonScripts/Scripts/GridFieldSetup/GridFieldSetup.py (hunk 1 above)
Comment (johnnywilkes) [reply Packs/CommonScripts/Scripts/GridFieldSetup/GridFieldSetup.py:26]: @esharf , why did you reverse the order?

File: Packs/CommonScripts/Scripts/GridFieldSetup/GridFieldSetup.py [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+
+    return res_list
+
+
+''' COMMAND FUNCTION '''
+
+
+def grid_field_setup_command(args: Dict[str, Any]) -> CommandResults:
+    keys = args.get('keys')
+    context_path = args.get('context_path')
```
[the remark this answers is not in the document]
Comment (johnnywilkes) [reply Packs/CommonScripts/Scripts/GridFieldSetup/GridFieldSetup.py:37]: I dont understand why this is needed or what you are proposing

File: Packs/CommonScripts/Scripts/GridFieldSetup/GridFieldSetup.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (johnnywilkes) [reply Packs/CommonScripts/Scripts/GridFieldSetup/GridFieldSetup.py:44]: how are these different?

File: Packs/CommonScripts/Scripts/GridFieldSetup/GridFieldSetup.py (hunk 1 above)
Comment (johnnywilkes) [reply Packs/CommonScripts/Scripts/GridFieldSetup/GridFieldSetup.py:26]: fixed in [switch to argToList](https://github.com/demisto/content/pull/21641/commits/3c29bbb2b836a8b3243fbeb9368bedea4d44399b)

File: Packs/CommonScripts/Scripts/GridFieldSetup/GridFieldSetup.py (hunk 2 above)
Comment (johnnywilkes) [reply Packs/CommonScripts/Scripts/GridFieldSetup/GridFieldSetup.py:20]: the structure of a grid field is a list of dictionaries

File: Packs/CommonScripts/Scripts/GridFieldSetup/README.md [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+| keys | columns for the grid field in comma separated format |
+| val1 | value for 1st key \(can be string or context path\) |
+| val2 | value for 2nd key \(can be string or context path\) |
+| val3 | value for 3rd key \(can be string or context path\) |
+| val4 | value for 4th key \(can be string or context path\) |
+| val5 | value for 5th key \(can be string or context path\) |
+
+## Outputs
+---
+There are no outputs for this script.
```
[the remark this answers is not in the document]
Comment (johnnywilkes) [reply Packs/CommonScripts/Scripts/GridFieldSetup/README.md:27]: @esharf , Does `demisto-sdk generate-docs -e` work with automations?

[the remark this answers is not in the document]
Comment (johnnywilkes) [reply Packs/CommonScripts/Scripts/GridFieldSetup/README.md:29]: this is complete

File: Packs/CommonScripts/Scripts/GridFieldSetup/GridFieldSetup_test.py [hunk 7]
```diff
@@ -0,0 +1,8 @@
+from GridFieldSetup import *
+
+
+def test_command(mocker):
```
[the remark this answers is not in the document]
Comment (johnnywilkes) [reply Packs/CommonScripts/Scripts/GridFieldSetup/GridFieldSetup_test.py]: this is complete

**Review by esharf — APPROVED** [review]

**Pull request merged**